### PR TITLE
topology: sof-adl-max98360a-rt5682-4ch: add new topology

### PR DESCRIPTION
**Pull request by brentlu**
Add support for four max98360a speaker amplifiers running in TDM mode
which format is 8 slots with 32 bit slot/sample width on ADL boards.

This topology implements a 4-channel pipeline directly to speaker
amplifiers so audio effects need to be done in user space.

Signed-off-by: Brent Lu <brent.lu@intel.com>

### CODE REVIEW

File: tools/topology/topology1/sof-tgl-max98357a-rt5682.m4 [hunk 1]
```diff
@@ -21,6 +21,11 @@ include(`platform/intel/'PLATFORM`.m4')
 include(`platform/intel/dmic.m4')
 DEBUG_START
 
+#
+# Check option conflicts
+#
+ifdef(`2_WAY_SPK', `ifdef(`4_CH_SPK', `fatal_error(note: 2_WAY_SPK and 4_CH_SPK are mutually exclusive)')')
```
Comment (lgirdwood) [inline tools/topology/topology1/sof-tgl-max98357a-rt5682.m4:27]: Probably best to use the same naming convention for both 2 and 4 CH speaker. 
In fact we could probably do a little better and fix ```2_WAY_SPK``` to ```SPK_CHANNELS_2``` and ```4_CHAN_SPK``` to ```SPK_CHANNELS_4``` in another patch in this PR.

Comment (uhs987) [reply tools/topology/topology1/sof-tgl-max98357a-rt5682.m4:27]: How about 4CH_2WAY and 4CH_PASSTHROUGH? First one means it's a four channel 2-way speaker design and second one is just four channel passthrough (from linux kernel side to ssp port).

Maybe we will have 6CH_3WAY in the future...

Comment (brentlu) [reply tools/topology/topology1/sof-tgl-max98357a-rt5682.m4:27]: How about 4CH_2WAY and 4CH_PASSTHROUGH? First one means it's a four channel 2-way speaker design and second one is just four channel passthrough (from linux kernel side to ssp port).

Maybe we will have 6CH_3WAY in the future...

Comment (plbossart) [reply tools/topology/topology1/sof-tgl-max98357a-rt5682.m4:27]: It's really confusing when we mix channels and transducers.

It's my understanding that two-way systems have woofer and tweeter transducers. There could be more than one transducer of each type, and that number would define the 'channel' count.

In other words, a L/R stream could be split into LwLt  and RwRt. That doesn't mean 4 channels to me.

**Comment by gkbldcig** [comment]
Can one of the admins verify this patch?

File: tools/topology/topology1/sof-tgl-max98357a-rt5682.m4 (hunk 1 above)
Comment (brentlu) [reply tools/topology/topology1/sof-tgl-max98357a-rt5682.m4:27]: @plbossart so do you think 2_WAY_SPK and 4_CH_SPK is good enough for you?

**Comment by lgirdwood** [comment]
please test this

**Comment by brentlu** [comment]
According to ODM's testing, the topology works on the test device.

File: tools/topology/topology1/sof-tgl-max98357a-rt5682.m4 (hunk 1 above)
Comment (plbossart) [reply tools/topology/topology1/sof-tgl-max98357a-rt5682.m4:27]: not really. 2-way speaker tells you a split of transducers, but not how many sets there are.
You could have 1, 2 or 4 sets of 2-way speakers.
likewise if there are 3-way speakers we could have 1, 2, 4 sets of 3-ways speakers.

Comment (brentlu) [reply tools/topology/topology1/sof-tgl-max98357a-rt5682.m4:27]: Is this case a 2-set 2-way speakers? Sorry I'm not familiar with the definition here.

Comment (plbossart) [reply tools/topology/topology1/sof-tgl-max98357a-rt5682.m4:27]: you could have L/R 2-way systems, or FL/FR/SL/SR 2-way systems. in the first case you would play 2 channels that would be split in 2 signals with 2 bands each.

**Comment by lgirdwood** [comment]
> According to ODM's testing, the topology works on the test device.

great - we just need to make updates requested by @plbossart and we are good !

File: tools/topology/topology1/sof-tgl-max98357a-rt5682.m4 (hunk 1 above)
Comment (brentlu) [reply tools/topology/topology1/sof-tgl-max98357a-rt5682.m4:27]: Rename macros like this?
2_WAY_SPK=>2CH_2WAY
4_CH_SPK => 4CH_PASSTHROUGH

Comment (plbossart) [reply tools/topology/topology1/sof-tgl-max98357a-rt5682.m4:27]: sounds good to me @brentlu

**Comment by brentlu** [comment]
rename the macro and check the tplg files are the same

**Comment by brentlu** [comment]
rebase for conflict

**Review by lgirdwood — APPROVED** [review]

**Comment by sathyap-chrome** [comment]
@brentlu kindly raise PR for merge on adl004

**Pull request merged**